Add 'expand' flag to the web plugin

maxammann opened the pull request:
Hello,

I'm currently developing an android app to synchronize your beets library with your android phone. This addition helps me to implement the query and simply some stuff.

There are some screenshots of my app attached.

![screenshot_20160613-140950](https://cloud.githubusercontent.com/assets/905221/16007099/dbc202c6-3170-11e6-9e01-8948204e844f.png)
![screenshot_20160613-140956](https://cloud.githubusercontent.com/assets/905221/16007098/dbc057d2-3170-11e6-980b-ff00039e7d63.png)

jackwilsdon commented on beetsplug/web/__init__.py:
```diff
@@ -55,7 +55,7 @@ def _rep(obj, expand=False):
 
 
-def json_generator(items, root):
+def json_generator(items, root, expand=False):
```
Do you think you could document this in the docstring for the method too?

jackwilsdon commented on beetsplug/web/__init__.py:
```diff
@@ -310,6 +314,9 @@ def func(lib, opts, args):
 
             app.config['lib'] = lib
+            # Normalizes json output
+            app.config['JSONIFY_PRETTYPRINT_REGULAR'] = False
```
Is there a need for this in production? From the Flask docs:

> If this is set to `True` (the default) jsonify responses will be pretty printed if they are not requested by an XMLHttpRequest object (controlled by the `X-Requested-With` header).

maxammann replied:
After experiencing with the json output of jsonify and json.dumps I noticed that the output of json.dumps is always compact while jsonify outputs it prettified.

That's the first problem, the second is that a prettified json string takes mus more space. Even the compat string of my library is already > 30 MB.

jackwilsdon replied:
Oh, oops! I read it as `True` instead of `False`, haha. Sorry about that and I agree with your reasoning 100%.

jackwilsdon commented:
Looks like our [flake8](https://travis-ci.org/beetbox/beets/jobs/137225888) test is failing due to some lines being more than 79 characters in length, but apart from that (and the comments I previously made) this looks great! Thanks :tada:.

jackwilsdon commented on beetsplug/web/__init__.py:
```diff
@@ -69,10 +69,14 @@ def json_generator(items, root):
     yield ']}'
 
 
+def is_expand(args):
```
Do we really need this method? It might be simpler to just do this:

```python
return flask.jsonify(_rep(entities[0], expand='expand' in flask.request.args))
```

Or even work out whether or not we should expand at the top of the `resource` function and use that instead?

maxammann replied:
Not sure, if I remove the method I would need to use the "in" operator

`
return flask.jsonify(_rep(entities[0], expand='expand' in flask.request.args))
`

This would also cause duplicate code. What do you think?

jackwilsdon replied:
Hmm, we'd be better with @sampsyo's opinion on this I think.

Here's how I'd (personally) write the method:

```python
def is_expand():
    """Returns whether the current request is for an expanded response."""

    return flask.request.args.get('expand') is not None
```

We can directly access `flask.request.args` without passing it in as flask does some magic with the `flask.request` object per-request so that it is the current request.

maxammann commented:
Thanks, added the recommendations

maxammann commented:
Is there a reson for line 40: `del out['path']`

The path variable would be very useful as I do not know the extension of the individual files and I'm not sure how the "format" tag is defined.

Thanks in advance!

jackwilsdon commented:
Hmm, looks like that was added by @sampsyo in 8493909a81a8feec1e83fe0fa99e4dd90a5c09a4. Not entirely sure why it's there, maybe we'd be better not giving the "full" path but instead the path relative to the beets library?

maxammann commented:
Oh yes, didn't notice that it returns the full path. How about replacing it with path.destination(fragment=True) instead of deleting it?

sampsyo commented:
Cool! This looks like a great project. Is it open-source also?

In the future, I'd love to talk to you about basing this kind of thing on the new AURA API we're developing: https://github.com/beetbox/aura

This API, when we've got it implemented, will include native support for stuff like this via "includes," which will let clients request exactly what data they want in their response. It would be great to collaborate on this if you're interested.

maxammann commented:
Yeah of course! FOSS FTW! :P
https://github.com/maxammann/music-cyclon

Started the rewrite a few days ago when I realized I could just integrate it directly into beets :smile: 
I also wanted to follow the material design guide lines.
Also a better api would be gold for music-cyclon as the unspecific requests are very slow and take a lot of time.

The synchronize part is the first milestone. I'm not sure yet whether streaming will be a thing or maybe remotly managing the library. For now synchronizing with poweramp integration is enough for me. Let me know what you think!

Looking forward to talk per chat or maybe even VOIP.

jackwilsdon commented:
You could join the beets IRC channel on freenode :smile:. You can join here: [freenode webchat](https://webchat.freenode.net/?channels=beets).

jackwilsdon commented:
Wow @sampsyo, I can't believe I've never seen aura! It looks very fleshed out and well designed :smile:. What's the long term plan for beets and aura? Create an aura plugin for beets?

sampsyo commented:
Yeah, that's the plan! The hope is to start with a new beets plugin that fits the API and then work on other parts of the ecosystem.

jrobeson commented on docs/plugins/web.rst:
```diff
@@ -197,6 +197,7 @@ The interface and response format is similar to the item API, except replacing
 or ``/album/5,7``. In addition we can request the cover art of an album with
 ``GET /album/5/art``.
+You can also add the '?expand' flag to get the individual items of an album.
 
```
can you provide an example of the usage of ?expand here?

sampsyo commented:
This looks ready to merge as soon as we have a changelog entry.

maxammann commented:
@sampsyo 
Web plugin Changelog:

- Added an option to show the items of an album
- Added the 'path' tag to the json outpu of a file which shows the relative path to the file
- Normalized the json output

sampsyo commented:
Could you please add that to `changelog.rst`? Then somebody can hit that green button.

maxammann commented:
Updated fork and added changelog

sampsyo commented:
Awesome; thank you! :sparkles:

maxammann commented:
Thanks as well :P
Do you know when the next release is going to be published?

sampsyo commented:
We don't have a precise release schedule, usually, but please feel free to weigh in if you think a faster release would ever be useful.
